File: src/OpStream.Client.Transports.WebSockets/WebSocketOpStreamClient.cs
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Collections.Concurrent;
using Microsoft.Extensions.Options;
using OpStream.Shared.Messages;

namespace OpStream.Client.Transports.WebSockets;

/// <summary>
/// A WebSocket-based implementation of the OpStream client transport.
/// Handles bidirectional communication using JSON messages over a single WebSocket.
/// </summary>
public class WebSocketOpStreamClient : IOpStreamClient
{
    private readonly ClientWebSocket _webSocket = new();
    private readonly ConcurrentDictionary<string, TaskCompletionSource<WebSocketMessage>> _pendingRequests = new();
    private readonly CancellationTokenSource _cts = new();
    private readonly Task _listenTask;
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
    };

    /// <inheritdoc/>
    public event Func<ReadOnlyMemory<byte>, long, Task>? OnReceiveOp;
    
    /// <inheritdoc/>
    public event Action<Exception?>? OnDisconnected;
    
    /// <inheritdoc/>
    public event Func<IEnumerable<AwarenessState>, Task>? OnReceiveAwareness;

    /// <inheritdoc/>
    public event Action<string>? OnPeerDisconnected;

    // ─── Comment events ───────────────────────────────────────────────────────

    private event Func<CommentDto, Task>? _onCommentCreated;
    private event Func<CommentDto, Task>? _onCommentUpdated;
    private event Func<CommentDeletedDto, Task>? _onCommentDeleted;

    /// <inheritdoc/>
    public event Func<CommentDto, Task>? OnCommentCreated
    {
        add    => _onCommentCreated += value;
        remove => _onCommentCreated -= value;
    }

    /// <inheritdoc/>
    public event Func<CommentDto, Task>? OnCommentUpdated
    {
        add    => _onCommentUpdated += value;
        remove => _onCommentUpdated -= value;
    }

    /// <inheritdoc/>
    public event Func<CommentDeletedDto, Task>? OnCommentDeleted
    {
        add    => _onCommentDeleted += value;
        remove => _onCommentDeleted -= value;
    }

    /// <summary>
    /// Initializes a new instance of the WebSocketOpStreamClient.
    /// </summary>
    /// <param name="options">Configuration options containing the server URI.</param>
    public WebSocketOpStreamClient(IOptions<OpStreamWebSocketOptions> options)
    {
        _listenTask = ListenLoop(new Uri(options.Value.ServerUri));
    }

    private async Task ListenLoop(Uri uri)
    {
        try
        {
            await _webSocket.ConnectAsync(uri, _cts.Token);

            while (_webSocket.State == WebSocketState.Open && !_cts.Token.IsCancellationRequested)
            {
                var json = await ReceiveTextAsync(_webSocket, _cts.Token);
                if (json == null) break;

                var message = JsonSerializer.Deserialize<WebSocketMessage>(json, JsonOptions);
                if (message == null) continue;

                if (!string.IsNullOrEmpty(message.CorrelationId) && _pendingRequests.TryRemove(message.CorrelationId, out var tcs))
                {
                    tcs.SetResult(message);
                    continue;
                }

                // Handle events
                switch (message.MessageType)
                {
                    case WebSocketOpMessageType.ReceiveOpEvent:
                        if (OnReceiveOp != null && message.ReceiveOpEvent != null)
                        {
                            await OnReceiveOp.Invoke(message.ReceiveOpEvent.Payload, message.ReceiveOpEvent.NewRevision);
                        }
                        break;
                    case WebSocketOpMessageType.ReceiveAwarenessEvent:
                        if (OnReceiveAwareness != null && message.ReceiveAwarenessEvent != null)
                        {
                            await OnReceiveAwareness.Invoke(message.ReceiveAwarenessEvent.Awareness);
                        }
                        break;
                    case WebSocketOpMessageType.PeerDisconnectedEvent:
                        if (message.PeerDisconnectedEvent != null)
                        {
                            OnPeerDisconnected?.Invoke(message.PeerDisconnectedEvent.PeerId);
                        }
                        break;
                    case WebSocketOpMessageType.ReceiveCommentCreated:
                        if (_onCommentCreated != null && message.ReceiveCommentCreated is JsonElement createdEl)
                        {
                            var dto = createdEl.Deserialize<CommentDto>(JsonOptions);
                            if (dto != null) await _onCommentCreated.Invoke(dto);
                        }
                        break;
                    case WebSocketOpMessageType.ReceiveCommentUpdated:
                        if (_onCommentUpdated != null && message.ReceiveCommentUpdated is JsonElement updatedEl)
                        {
                            var dto = updatedEl.Deserialize<CommentDto>(JsonOptions);
                            if (dto != null) await _onCommentUpdated.Invoke(dto);
                        }
                        break;
                    case WebSocketOpMessageType.ReceiveCommentDeleted:
                        if (_onCommentDeleted != null && message.ReceiveCommentDeleted != null)
                        {
                            await _onCommentDeleted.Invoke(new CommentDeletedDto(message.ReceiveCommentDeleted.CommentId));
                        }
                        break;
                }
            }
        }
        catch (OperationCanceledException) { }
        catch (Exception ex)
        {
            OnDisconnected?.Invoke(ex);
        }
        finally
        {
            if (_webSocket.State == WebSocketState.Open || _webSocket.State == WebSocketState.CloseReceived)
            {
                try { await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Closing", CancellationToken.None); } catch { }
            }
        }
    }

    private async Task<string?> ReceiveTextAsync(WebSocket webSocket, CancellationToken ct)
    {
        var buffer = new byte[1024 * 4];
        using var ms = new MemoryStream();
        
        WebSocketReceiveResult result;
        do
        {
            result = await webSocket.ReceiveAsync(new ArraySegment<byte>(buffer), ct);
            if (result.MessageType == WebSocketMessageType.Close) return null;
            ms.Write(buffer, 0, result.Count);
        } while (!result.EndOfMessage);

        ms.Seek(0, SeekOrigin.Begin);
        using var reader = new StreamReader(ms, Encoding.UTF8);
        return await reader.ReadToEndAsync();
    }

    /// <inheritdoc/>
    public async Task<ClientJoinResult> ConnectAndJoinAsync(string documentId, string documentType, CancellationToken ct = default)
    {
        // Wait for connection to be established in ListenLoop
        while (_webSocket.State == WebSocketState.None || _webSocket.State == WebSocketState.Connecting)
        {
            await Task.Delay(10, ct);
        }

        if (_webSocket.State != WebSocketState.Open)
        {
            throw new Exception("WebSocket is not open");
        }

        var correlationId = Guid.NewGuid().ToString();
        var tcs = new TaskCompletionSource<WebSocketMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pendingRequests[correlationId] = tcs;

        var request = new WebSocketMessage
        {
            CorrelationId = correlationId,
            MessageType = WebSocketOpMessageType.JoinRequest,
            JoinRequest = new JoinRequestData(documentId, documentType, 1)
        };

        await SendMessageAsync(request, ct);

        using var registration = ct.Register(() => tcs.TrySetCanceled());
        var response = await tcs.Task;

        if (response.MessageType == WebSocketOpMessageType.ErrorResponse)
        {
            throw new Exception(response.ErrorMessage);
        }

        if (response.MessageType != WebSocketOpMessageType.JoinResponse || response.JoinResponse == null)
        {
            throw new Exception("Unexpected response during Join");
        }

        var jr = response.JoinResponse;
        return new ClientJoinResult(jr.Revision, jr.Snapshot, jr.Awareness);
    }

    /// <inheritdoc/>
    public async Task<ClientOpResult> SendOpAsync(string documentId, ReadOnlyMemory<byte> payload, long baseRevision, CancellationToken ct = default)
    {
        var correlationId = Guid.NewGuid().ToString();
        var tcs = new TaskCompletionSource<WebSocketMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pendingRequests[correlationId] = tcs;

        var request = new WebSocketMessage
        {
            CorrelationId = correlationId,
            MessageType = WebSocketOpMessageType.OpRequest,
            OpRequest = new OpRequestData(documentId, payload.ToArray(), baseRevision)
        };

        await SendMessageAsync(request, ct);

        using var registration = ct.Register(() => tcs.TrySetCanceled());
        var response = await tcs.Task;

        if (response.MessageType == WebSocketOpMessageType.ErrorResponse)
        {
            throw new Exception(response.ErrorMessage);
        }

        if (response.MessageType != WebSocketOpMessageType.OpResponse || response.OpResponse == null)
        {
            throw new Exception("Unexpected response during SendOp");
        }

        var or = response.OpResponse;
        return new ClientOpResult(or.Success, or.NewRevision, or.ErrorMessage);
    }

    /// <inheritdoc/>
    public async Task SendAwarenessAsync(string documentId, JsonElement data, CancellationToken ct = default)
    {
        var request = new WebSocketMessage
        {
            MessageType = WebSocketOpMessageType.AwarenessRequest,
            AwarenessRequest = new AwarenessRequestData(documentId, data.GetRawText())
        };

        await SendMessageAsync(request, ct);
    }

    // ─── Comments ────────────────────────────────────────────────────────────

    /// <inheritdoc/>
    public async Task<List<CommentDto>> ListOpenCommentsAsync(string documentId, CancellationToken ct = default)
    {
        var response = await SendCommentRequestAsync(
            WebSocketOpMessageType.ListOpenComments,
            new CommentCommandData { DocumentId = documentId },
            ct);

        if (response.CommentResponse is not JsonElement el)
            throw new Exception("Unexpected response for ListOpenComments.");
        return el.Deserialize<List<CommentDto>>(JsonOptions) ?? new List<CommentDto>();
    }

    /// <inheritdoc/>
    public async Task<CommentDto> CreateCommentAsync(string documentId, NewCommentCmd cmd, CancellationToken ct = default)
    {
        JsonElement? anchorEl = cmd.Anchor is null
            ? null
            : JsonSerializer.SerializeToElement(cmd.Anchor, JsonOptions);

        var response = await SendCommentRequestAsync(
            WebSocketOpMessageType.CreateComment,
            new CommentCommandData
            {
                DocumentId = documentId,
                Body = cmd.Body,
                ParentCommentId = cmd.ParentCommentId,
                Anchor = anchorEl
            }, ct);

        return DeserializeCommentResponse(response);
    }

    /// <inheritdoc/>
    public async Task<CommentDto> EditCommentAsync(string documentId, string commentId, string newBody, CancellationToken ct = default)
    {
        var response = await SendCommentRequestAsync(
            WebSocketOpMessageType.EditComment,
            new CommentCommandData { DocumentId = documentId, CommentId = commentId, Body = newBody },
            ct);

        return DeserializeCommentResponse(response);
    }

    /// <inheritdoc/>
    public async Task<CommentDto> ResolveCommentAsync(string documentId, string commentId, CancellationToken ct = default)
    {
        var response = await SendCommentRequestAsync(
            WebSocketOpMessageType.ResolveComment,
            new CommentCommandData { DocumentId = documentId, CommentId = commentId },
            ct);

        return DeserializeCommentResponse(response);
    }

    /// <inheritdoc/>
    public async Task DeleteCommentAsync(string documentId, string commentId, CancellationToken ct = default)
    {
        await SendCommentRequestAsync(
            WebSocketOpMessageType.DeleteComment,
            new CommentCommandData { DocumentId = documentId, CommentId = commentId },
            ct);
    }

    private async Task<WebSocketMessage> SendCommentRequestAsync(
        WebSocketOpMessageType messageType, CommentCommandData cmd, CancellationToken ct)
    {
        var correlationId = Guid.NewGuid().ToString();
        var tcs = new TaskCompletionSource<WebSocketMessage>(TaskCreationOptions.RunContinuationsAsynchronously);
        _pendingRequests[correlationId] = tcs;

        await SendMessageAsync(new WebSocketMessage
        {
            CorrelationId = correlationId,
            MessageType = messageType,
            CommentCommand = cmd
        }, ct);

        using var registration = ct.Register(() => tcs.TrySetCanceled());
        var response = await tcs.Task;

        if (response.MessageType == WebSocketOpMessageType.ErrorResponse)
            throw new Exception(response.ErrorMessage ?? "Comment operation failed.");

        return response;
    }

    private static CommentDto DeserializeCommentResponse(WebSocketMessage response)
    {
        if (response.CommentResponse is not JsonElement el)
            throw new Exception("Unexpected response format for comment operation.");
        return el.Deserialize<CommentDto>(JsonOptions)
            ?? throw new Exception("Null comment in server response.");
    }

    private async Task SendMessageAsync(WebSocketMessage message, CancellationToken ct)
    {
        var json = JsonSerializer.Serialize(message, JsonOptions);
        var bytes = Encoding.UTF8.GetBytes(json);
        await _webSocket.SendAsync(new ArraySegment<byte>(bytes), WebSocketMessageType.Text, true, ct);
    }

    /// <inheritdoc/>
    public async ValueTask DisposeAsync()
    {
        _cts.Cancel();
        if (_webSocket.State == WebSocketState.Open)
        {
            try { await _webSocket.CloseAsync(WebSocketCloseStatus.NormalClosure, "Disposing", CancellationToken.None); } catch { }
        }
        _webSocket.Dispose();
        try { await _listenTask; } catch { }
        _cts.Dispose();
    }
}

File: samples/OpStream.BrowserExtension/collab/WebSocketOpStreamClient.js
/**
 * A WebSocket-based implementation of the OpStream client transport for HTML + JS clients.
 * Handles bidirectional communication using JSON messages over a single WebSocket.
 *
 * NOTE: copied verbatim from samples/MonacoCollaborativeJs/js/WebSocketOpStreamClient.js
 * so the browser extension is self-contained and packageable.
 */
export class WebSocketOpStreamClient {
    constructor(serverUri) {
        this.serverUri = serverUri;
        this._webSocket = null;
        this._pendingRequests = new Map();

        // ── Events ────────────────────────────────────────────────────────
        this.onReceiveOp = null;
        this.onDisconnected = null;
        this.onReceiveAwareness = null;
        this.onPeerDisconnected = null;

        this.onCommentCreated = null;
        this.onCommentUpdated = null;
        this.onCommentDeleted = null;

        // Maps to C# WebSocketOpMessageType enum
        this.MessageType = {
            JoinRequest: 0,
            JoinResponse: 1,
            OpRequest: 2,
            OpResponse: 3,
            AwarenessRequest: 4,
            ReceiveOpEvent: 5,
            ReceiveAwarenessEvent: 6,
            PeerDisconnectedEvent: 7,
            ErrorResponse: 8,
            CreateComment: 9,
            EditComment: 10,
            ResolveComment: 11,
            DeleteComment: 12,
            ListOpenComments: 13,
            ReceiveCommentCreated: 14,
            ReceiveCommentUpdated: 15,
            ReceiveCommentDeleted: 16
        };

        this._connect();
    }

    _connect() {
        if (typeof WebSocket === 'undefined') {
            throw new Error("WebSocket is not supported in this environment.");
        }

        this._webSocket = new WebSocket(this.serverUri);

        this._webSocket.onmessage = async (event) => {
            try {
                const message = JSON.parse(event.data);

                if (message.correlationId && this._pendingRequests.has(message.correlationId)) {
                    const pending = this._pendingRequests.get(message.correlationId);
                    this._pendingRequests.delete(message.correlationId);
                    pending.resolve(message);
                    return;
                }

                switch (message.messageType) {
                    case this.MessageType.ReceiveOpEvent:
                    case "ReceiveOpEvent":
                        if (this.onReceiveOp && message.receiveOpEvent) {
                            await this.onReceiveOp(message.receiveOpEvent.payload, message.receiveOpEvent.newRevision);
                        }
                        break;
                    case this.MessageType.ReceiveAwarenessEvent:
                    case "ReceiveAwarenessEvent":
                        if (this.onReceiveAwareness && message.receiveAwarenessEvent) {
                            await this.onReceiveAwareness(message.receiveAwarenessEvent.awareness);
                        }
                        break;
                    case this.MessageType.PeerDisconnectedEvent:
                    case "PeerDisconnectedEvent":
                        if (this.onPeerDisconnected && message.peerDisconnectedEvent) {
                            this.onPeerDisconnected(message.peerDisconnectedEvent.peerId);
                        }
                        break;
                    case this.MessageType.ReceiveCommentCreated:
                    case "ReceiveCommentCreated":
                        if (this.onCommentCreated && message.receiveCommentCreated) {
                            await this.onCommentCreated(message.receiveCommentCreated);
                        }
                        break;
                    case this.MessageType.ReceiveCommentUpdated:
                    case "ReceiveCommentUpdated":
                        if (this.onCommentUpdated && message.receiveCommentUpdated) {
                            await this.onCommentUpdated(message.receiveCommentUpdated);
                        }
                        break;
                    case this.MessageType.ReceiveCommentDeleted:
                    case "ReceiveCommentDeleted":
                        if (this.onCommentDeleted && message.receiveCommentDeleted) {
                            await this.onCommentDeleted({ deletedCommentId: message.receiveCommentDeleted.commentId });
                        }
                        break;
                }
            } catch (ex) {
                console.error("Failed to process WebSocket message", ex);
            }
        };

        this._webSocket.onclose = (event) => {
            if (this.onDisconnected) {
                this.onDisconnected(new Error(`WebSocket closed. Code: ${event.code}`));
            }
        };

        this._webSocket.onerror = () => {
            if (this.onDisconnected) {
                this.onDisconnected(new Error("WebSocket error occurred."));
            }
        };
    }

    async _waitForConnection() {
        if (!this._webSocket) return;
        if (this._webSocket.readyState === 1) return;

        return new Promise((resolve, reject) => {
            const check = () => {
                if (this._webSocket.readyState === 1) {
                    resolve();
                } else if (this._webSocket.readyState > 1) {
                    reject(new Error("WebSocket is not open"));
                } else {
                    setTimeout(check, 10);
                }
            };
            check();
        });
    }

    async connectAndJoinAsync(documentId, documentType) {
        await this._waitForConnection();

        if (this._webSocket.readyState !== 1) {
            throw new Error("WebSocket is not open");
        }

        const correlationId = this._generateId();
        const request = {
            correlationId: correlationId,
            messageType: this.MessageType.JoinRequest,
            joinRequest: {
                documentId: documentId,
                documentType: documentType,
                clientProtoVersion: 1
            }
        };

        const response = await this._sendMessageAsync(request, correlationId);

        if (response.messageType === this.MessageType.ErrorResponse || response.messageType === "ErrorResponse") {
            throw new Error(response.errorMessage || "Join failed.");
        }
        if (response.messageType !== this.MessageType.JoinResponse && response.messageType !== "JoinResponse") {
            throw new Error("Unexpected response during Join");
        }

        const jr = response.joinResponse;
        return {
            revision: jr.revision,
            snapshot: jr.snapshot,
            currentAwareness: jr.awareness || []
        };
    }

    async sendOpAsync(documentId, payload, baseRevision) {
        const correlationId = this._generateId();

        let encodedPayload = payload;
        if (payload instanceof Uint8Array || payload instanceof ArrayBuffer) {
            encodedPayload = this._arrayBufferToBase64(payload);
        }

        const request = {
            correlationId: correlationId,
            messageType: this.MessageType.OpRequest,
            opRequest: {
                documentId: documentId,
                payload: encodedPayload,
                baseRevision: baseRevision
            }
        };

        const response = await this._sendMessageAsync(request, correlationId);

        if (response.messageType === this.MessageType.ErrorResponse || response.messageType === "ErrorResponse") {
            throw new Error(response.errorMessage || "SendOp failed.");
        }
        if (response.messageType !== this.MessageType.OpResponse && response.messageType !== "OpResponse") {
            throw new Error("Unexpected response during SendOp");
        }

        const or = response.opResponse;
        return {
            success: or.success,
            newRevision: or.newRevision,
            errorMessage: or.errorMessage
        };
    }

    async sendAwarenessAsync(documentId, data) {
        const request = {
            messageType: this.MessageType.AwarenessRequest,
            awarenessRequest: {
                documentId: documentId,
                dataJson: typeof data === 'string' ? data : JSON.stringify(data)
            }
        };
        this._webSocket.send(JSON.stringify(request));
    }

    async _sendMessageAsync(message, correlationId) {
        return new Promise((resolve, reject) => {
            this._pendingRequests.set(correlationId, { resolve, reject });
            this._webSocket.send(JSON.stringify(message));
        });
    }

    async disposeAsync() {
        if (this._webSocket) {
            if (this._webSocket.readyState === 1 || this._webSocket.readyState === 0) {
                this._webSocket.close(1000, "Disposing");
            }
        }
        for (const pending of this._pendingRequests.values()) {
            pending.reject(new Error("Client disposed"));
        }
        this._pendingRequests.clear();
        return Promise.resolve();
    }

    _generateId() {
        return typeof crypto !== 'undefined' && crypto.randomUUID
            ? crypto.randomUUID()
            : Math.random().toString(36).substring(2, 15);
    }

    _arrayBufferToBase64(buffer) {
        let binary = '';
        const bytes = new Uint8Array(buffer);
        const len = bytes.byteLength;
        for (let i = 0; i < len; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return typeof btoa !== 'undefined' ? btoa(binary) : Buffer.from(buffer).toString('base64');
    }
}
